src/segtree.rs

```diff
@@ -1 +1,314 @@
+use crate::internal_bit::ceil_pow2;
+use crate::internal_type_traits::{BoundedAbove, BoundedBelow, One, Zero};
+use std::cmp::{max, min};
+use std::convert::Infallible;
+use std::marker::PhantomData;
+use std::ops::{Add, Mul};
 
+// TODO Should I split monoid-related traits to another module?
+pub trait Monoid {
+    type S: Clone;
+    fn identity() -> Self::S;
+    fn binary_operation(a: &Self::S, b: &Self::S) -> Self::S;
+}
+
+pub struct Max<S>(Infallible, PhantomData<fn() -> S>);
+impl<S> Monoid for Max<S>
+where
+    S: Copy + Ord + BoundedBelow,
+{
+    type S = S;
+    fn identity() -> Self::S {
+        S::min_value()
+    }
+    fn binary_operation(a: &Self::S, b: &Self::S) -> Self::S {
+        max(*a, *b)
+    }
+}
+
+pub struct Min<S>(Infallible, PhantomData<fn() -> S>);
+impl<S> Monoid for Min<S>
+where
+    S: Copy + Ord + BoundedAbove,
+{
+    type S = S;
+    fn identity() -> Self::S {
+        S::max_value()
+    }
+    fn binary_operation(a: &Self::S, b: &Self::S) -> Self::S {
+        min(*a, *b)
+    }
+}
+
+pub struct Sum<S>(Infallible, PhantomData<fn() -> S>);
+impl<S> Monoid for Sum<S>
+where
+    S: Copy + Add<Output = S> + Zero,
+{
+    type S = S;
+    fn identity() -> Self::S {
+        S::zero()
+    }
+    fn binary_operation(a: &Self::S, b: &Self::S) -> Self::S {
+        *a + *b
```

Comment:
We can write like this but maybe it's excessive...

```rust
pub struct Sum<S>(Infallible, PhantomData<fn() -> S>);
impl<S> Monoid for Sum<S>
where
    S: Copy + Zero,
    for<'a> &'a S: Add<&'a S, Output = S>,
{
    type S = S;
    fn identity() -> Self::S {
        S::zero()
    }
    fn binary_operation(a: &Self::S, b: &Self::S) -> Self::S {
        a + b
    }
}

const _: fn() = || {
    fn pass<S>()
    where
        Sum<S>: Monoid,
    {
    }

    pass::<i32>();
    pass::<usize>();
};
```

Reply:
You mean that since `+` operators are defined between references, we can change the trait bound so that it works for all the types that defines the add operator between its reference instead of the raw type itself?

Reply:
Yes, that's I meant. With this `Zero` definition,  for example `Sum<num_bigint::BigInt>: Monoid` will be automatically implemented. But as I said, it may be excessive.

```rust
use static_assertions::assert_impl_all;
use std::iter::{self, Sum};

pub trait Zero: Sum {
    fn zero() -> Self;
}

impl<T: Sum> Zero for T {
    #[inline]
    fn zero() -> Self {
        iter::empty().sum()
    }
}

assert_impl_all!(usize: Zero);
assert_impl_all!(i64: Zero);
assert_impl_all!(f64: Zero);
assert_impl_all!(num_bigint::BigInt: Zero);
assert_impl_all!(ac_library_rs::ModInt1000000007: Zero);
```

Reply:
In the original implementation, `op` does not take references. As we have discussed, we should not modify the interface if there is no special reason. So please forget this.